pkg/alertmanager/multitenant.go

```diff
@@ -244,11 +244,11 @@ type MultitenantAlertmanager struct {
 	// effect here.
 	fallbackConfig string
 
-	// All the organization configurations that we have. Only used for instrumentation.
-	cfgs map[string]alerts.AlertConfigDesc
-
 	alertmanagersMtx sync.Mutex
 	alertmanagers    map[string]*Alertmanager
+	// Stores the current set of configurations we're running in each tenant's Alertmanager.
+	// Used for comparing configurations as we synchronize them.
+	cfgs map[string]alerts.AlertConfigDesc
```

Comment:
I would expect `cfgs` to be protected by `alertmanagerMtx` (because there may be concurrent accesses to it) but its usage in `setConfig()` is not protected. We should fix it.

Reply:
Good shout. 👌

Reply:
I went for a lock at the top of the function as there are several accesses to its existence and the config setting throughout the function, as opposed to lock/unlock multiple times throughout the method. This feels simpler than trying to optimise in multiple accesses for now but let me know what you think!

Reply:
I don't think we should lock the function. There are some blocking calls down there, while we should keep the lock the least as possible (considering it's an exclusive global lock). There are only two usages of `am.cfgs` in `setConfig()` so doesn't look prohibitive took a lock at the right place.

In particular, to my understanding of the function logic, `am.cfgs[cfg.User]` (used in the `if`) can be read along with the following:

```
am.alertmanagersMtx.Lock()
existing, hasExisting := am.alertmanagers[cfg.User]
am.alertmanagersMtx.Unlock()
```

What do you think?

Reply:
Generally, I'd yes, but if we look at the problem holistically - I think it makes sense to just lock the whole function.

First, the only way we can call `setConfig` concurrently is if we have reached the poll interval or there was a change of topology in the ring. On the latter case, there will certainly be configs that are going to be removed/added from a particular alertmanager.

Second, as you're already aware `setConfig` operates on the assumption that the alertmanager exists or not for that particular tenant. So the moment we do:
```
am.alertmanagersMtx.Lock()
existing, hasExisting := am.alertmanagers[cfg.User]
am.alertmanagersMtx.Unlock()
```

We're running the function under the assumption that the tenant exists for this particular AM but that's not necessarily true. If we release the lock another run of `syncConfigs` could have removed it by the time we do the work to create/update the alertmanager. 

I understand the chances of this race are fairly slim, but I genuinely think this is the safest route. Is there anything I'm missing?

Reply:
> First, the only way we can call setConfig concurrently is if we have reached the poll interval or there was a change of topology in the ring.

`setConfig()` can also be called by `alertmanagerFromFallbackConfig()`, which can be called while serving an HTTP request from the user.

> We're running the function under the assumption that the tenant exists for this particular AM but that's not necessarily true. If we release the lock another run of syncConfigs could have removed it by the time we do the work to create/update the alertmanager.

This is true. `loadAndSyncConfigs()` cannot be called concurrently, but `setConfig()` can be both called concurrently by `loadAndSyncConfigs()` (its calls chain) and an HTTP request setting up the alertmanager with the default config.

LGTM 👍